Add article about Stores in Message Queue (#4882)
magento/devdocs: guides/v2.3/extension-dev-guide/message-queues/message-queues-stores.md

            if ($headers instanceof AMQPTable) {
                $headers = $headers->getNativeData();
            }
            if (isset($headers['store_id'])) {
                $storeId = $headers['store_id'];
                try {
                    $currentStoreId = $this->storeManager->getStore()->getId();
                } catch (NoSuchEntityException $e) {
                    $this->logger->error(
                        sprintf(
                            "Can't set currentStoreId during processing queue. Message rejected. Error %s.",
                            $e->getMessage()
                        )
                    );
                    $subject->getQueue()->reject($message, false, $e->getMessage());
                    return;
                }
                if (isset($storeId) && $storeId !== $currentStoreId) {
                    $this->storeManager->setCurrentStore($storeId);
                }
            }
        }
        $proceed($message);
        if (isset($storeId, $currentStoreId) && $storeId !== $currentStoreId) {
            $this->storeManager->setCurrentStore($currentStoreId);//restore original store value
        }
    }
```

> The plugin checks the message headers and set the current store value in `storeManager` to the received `store_id` value.

Reviewer: ```suggestion
The plugin checks the message headers and sets the current store value in `storeManager` to the received `store_id` value.
```

---
group: php-developer-guide

> title: Architecture of Message Queues

Reviewer: ```suggestion
title: Message queues and stores
```

contributor_name: Comwrap GmbH
contributor_link: https://www.comwrap.com
functional_areas:
  - Services
---

One of the key features of Message Queues - is a possibility of processing all asynchronous requests only for particular stores.

For support of store scopes extension `Magento_AmqpStore` is responsible. 

{: .bs-callout .bs-callout-warning }

> Support of stores and mentioned extension is only available starting from Magento 2.3.3 version. If you are using version 2.3.1 or 2.3.2, then you can use Magento [PATCH]:https://magento.com/tech-resources/download#download2312

Author: Should we remove this note then in version of documentation for 2.3.3?

                    $this->logger->error($errorMessage);
                    throw new AMQPInvalidArgumentException($errorMessage);
                }
                $properties['application_headers'] = $headers;
            }
        } else {
            $properties['application_headers'] = new AMQPTable(['store_id' => $storeId]);
        }
        $updatedEnvelopes[] = $this->envelopeFactory->create(
            [
                'body' => $envelope->getBody(),
                'properties' => $properties
            ]
        );
    }
    if (!empty($updatedEnvelopes)) {
        $envelopes = $updatedEnvelopes;
    }
    return [$topic, $envelopes];
}
```

From example before you can see that plugin checks `application_headers` and add there `store_id` parameter. If headers are not exist, then plugin will create them. In this way RabbitMQ message will receive information for which store requrest have to be processed.

### Processing by consumer

[Consumers]:https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-queue.html are picking up messages from RabbitMQ queue and processing them. 

> So on a step when consumer reads message, extension executes around plugin in: 

Reviewer: ```suggestion
On a step when a consumer reads a message, the extension executes an around plugin, as shown here: 
```

            if ($headers instanceof AMQPTable) {
                $headers = $headers->getNativeData();
            }
            if (isset($headers['store_id'])) {
                $storeId = $headers['store_id'];
                try {
                    $currentStoreId = $this->storeManager->getStore()->getId();
                } catch (NoSuchEntityException $e) {
                    $this->logger->error(
                        sprintf(
                            "Can't set currentStoreId during processing queue. Message rejected. Error %s.",
                            $e->getMessage()
                        )
                    );
                    $subject->getQueue()->reject($message, false, $e->getMessage());
                    return;
                }
                if (isset($storeId) && $storeId !== $currentStoreId) {
                    $this->storeManager->setCurrentStore($storeId);
                }
            }
        }
        $proceed($message);
        if (isset($storeId, $currentStoreId) && $storeId !== $currentStoreId) {
            $this->storeManager->setCurrentStore($currentStoreId);//restore original store value
        }
    }
```

> Plugin will check message headers and will set up current store in `storeManager` to received `store_id` value.

Reviewer: ```suggestion
The plugin checks the message headers and set the current store value in `storeManager` to the received `store_id` value.
```
